### PR TITLE
Store shard assignments as bitsets

### PR DESCRIPTION
**Pull request by mmihic**
Required by a downstream PR to build the mapping rule sets that are actually used during query fan-out.

R: @xichen2020 @robskillington

### CODE REVIEW

File: .ci/auto-gen.sh [hunk 1]
```diff
@@ -102,6 +114,8 @@ set -e
 
 if [ "$2" = "generated/mocks" ]; then
     mocks_clear
+elif [ "$s" = "generated/proto" ]; then
```
Comment (xichen2020) [inline .ci/auto-gen.sh:117]: Probably should be "$2"?

File: placement/placement.go [hunk 2]
```diff
@@ -31,6 +31,7 @@ import (
 	"github.com/m3db/m3storage/generated/proto/schema"
 	"github.com/m3db/m3x/log"
 	"github.com/m3db/m3x/time"
+	"github.com/willf/bitset"
```
Comment (xichen2020) [inline placement/placement.go:34]: Add a blank line between this and the import path above?

File: placement/placement.go [hunk 3]
```diff
@@ -432,49 +433,58 @@ func (sp storagePlacement) commitDatabaseChanges(
...
+			removed.Set(i)
+			numToRemove--
+		}
 
-	// Sort the unowned shards so that assignments are deterministic (for
-	// testing)
-	sort.Sort(shardsInOrder(unowned))
+		unowned.InPlaceUnion(removed)
+		shardCutoffs[name] = removed
+		db.ShardAssignments[name].Bits = assigned.Difference(removed).Bytes()
```
Comment (xichen2020) [inline placement/placement.go:458]: nit: use `InPlaceDifference` to save a copy

File: placement/placement.go [hunk 4]
```diff
@@ -432,49 +433,58 @@ func (sp storagePlacement) commitDatabaseChanges(
...
+		for i, e := unowned.NextSet(0); e && numToAdd > 0; i, e = unowned.NextSet(i + 1) {
+			sp.log.Infof("assigned shard %d to %s:%s", i, db.Properties.Name, name)
+			added.Set(i)
+			numToAdd--
+		}
 
-		// Sort the shards so if we unassign later, the results will be deterministic
-		sort.Sort(shardsInOrder(assigned.Shards))
+		unowned.InPlaceDifference(added)
+		db.ShardAssignments[name].Bits = assigned.Union(added).Bytes()
```
Comment (xichen2020) [inline placement/placement.go:486]: nit: use `InPlaceUnion` to save a copy

File: placement/placement.go [hunk 5]
```diff
@@ -432,49 +433,58 @@ func (sp storagePlacement) commitDatabaseChanges(
...
+		shardCutovers[name] = added
 	}
 
 	// now build out all of the rules
```
Comment (xichen2020) [inline placement/placement.go:490]: Do we want to validate that `unowned` is empty after pass 2?

**Comment by mmihic** [comment]
I'm abandoning this branch; some of the functionality is getting rolled into a new branch which includes the building of the mapping chain.

**Pull request closed without merging**